source/MaterialXGraphEditor/Graph.cpp

```diff
@@ -3566,96 +3600,66 @@ void Graph::addNodePopup(bool cursor)
+
+         // Filter nodedefs and add to menu if matches filter
+        for (auto node : _nodesToAdd)
         {
             // Filter out list of nodes
             if (subs.size() > 0)
             {
                 ImGui::SetNextWindowSizeConstraints(ImVec2(250.0f, 300.0f), ImVec2(-1.0f, 500.0f));
-                for (size_t i = 0; i < it->second.size(); i++)
-                {
-                    std::string str(it->second[i][0]);
-                    std::string nodeName = it->second[i][0];
+                std::string str(node[0]);
+                std::string nodeName = node[0];
 
-                    // Disallow creating nested nodegraphs 
-                    if (_isNodeGraph && it->first == NODEGRAPH_ENTRY)
-                    {
-                        continue;
-                    }
+                // Disallow creating nested nodegraphs
+                if (_isNodeGraph && node[3] == NODEGRAPH_ENTRY)
+                {
+                    continue;
+                }
 
-                    // Allow spaces to be used to search for node names
-                    std::replace(subs.begin(), subs.end(), ' ', '_');
+                // Allow spaces to be used to search for node names
+                std::replace(subs.begin(), subs.end(), ' ', '_');
 
-                    if (str.find(subs) != std::string::npos)
+                if (str.find(subs) != std::string::npos)
+                {
+                    if (ImGui::MenuItem(getUserNodeDefName(nodeName).c_str()) || (ImGui::IsItemFocused() && ImGui::IsKeyPressedMap(ImGuiKey_Enter)))
                     {
-                        if (ImGui::MenuItem(getUserNodeDefName(nodeName).c_str()) || (ImGui::IsItemFocused() && ImGui::IsKeyPressedMap(ImGuiKey_Enter)))
-                        {
-                            addNode(it->second[i][2], getUserNodeDefName(nodeName), it->second[i][1]);
-                            _addNewNode = true;
-                            memset(input, '\0', sizeof(input));
-                        }
+                        addNode(node[2], getUserNodeDefName(nodeName), node[1]);
+                        _addNewNode = true;
+                        memset(input, '\0', sizeof(input));
                     }
                 }
             }
             else
             {
                 ImGui::SetNextWindowSizeConstraints(ImVec2(100, 10), ImVec2(-1, 300));
-                if (ImGui::BeginMenu(it->first.c_str()))
+                if (ImGui::BeginMenu(node[3].c_str()))
                 {
                     ImGui::SetWindowFontScale(_fontScale);
-                    for (size_t j = 0; j < it->second.size(); j++)
+                    std::string name = node[0];
+                    std::string prefix = "ND_";
+                    if (name.compare(0, prefix.size(), prefix) == 0 && name.compare(prefix.size(), std::string::npos, node[2]) == 0)
```

Comment:
Sorry. I'm not clear what this does ? Just checking as not all definitions start with "ND_".

Reply:
This is an attempt to codify the 'rule' for nesting menus from this slack comment (https://academysoftwarefdn.slack.com/archives/C0230LWBE2X/p1697132570854279?thread_ts=1697130472.487569&cid=C0230LWBE2X). In looking at the point_light node, the actual node name is `ND_point_light` but the category (or getNodeString()) returns `point_light`. To me the rule then is:

Wherever name == category (without the ND_ prefix) we just show that node in the 2nd layer, otherwise we add the category as a menu item, and put the name as the node in the 3rd layer.